libr/debug/p/native/linux/linux_debug.c

```diff
@@ -510,32 +549,30 @@ static bool linux_kill_thread(int tid, int signo) {
 	return true;
 }
 
-static bool linux_stop_thread(int tid) {
-	int status;
-	int ret = -1;
-
-	if (linux_kill_thread (tid, SIGSTOP)) {
-		if ((ret = waitpid (tid, &status, WUNTRACED)) == -1) {
-			perror("waitpid");
-		}
+static bool linux_stop_thread(RDebug *dbg, int tid) {
+	int ret;
+	siginfo_t siginfo = { 0 };
+	// Return if thread is already stopped to prevent an extra SIGSTOP signal
+	ret = r_debug_ptrace (dbg, PTRACE_GETSIGINFO, tid, 0, (r_ptrace_data_t)(intptr_t)&siginfo);
+	if (ret != -1) {
```

Comment:
Based on the ptrace manuals, it says that most of the ptrace commands will fail if the target is not stopped and I believe `PTRACE_GETSIGINFO` is included. I am not checking the content because of that and if it contains other signals, we can receive those signals later in the next `waitpid`.

Reply:
> will fail if the target is not stopped

So you will return from the function without stopping the thread in the current implementation.

Reply:
Checking signals with getsignals won't remove them, you still need to waitpid for that. Relying on this error sounds like undefined behaviour for any case except for checking if the debugger is attached. You can check errno to see if actually fails because the process is running.

Reply:
> So you will return from the function without stopping the thread in the current implementation.

That was my bad and I see the problem now. It should return when success and `PTRACE_GETSIGINFO` returns 0.

> Checking signals with getsignals won't remove them, you still need to waitpid for that. Relying on this error sounds like undefined behaviour for any case except for checking if the debugger is attached. You can check errno to see if it actually fails because the process is running.

I agree with you on this. `PTRACE_GETSIGINFO` will not remove the signals and I will need to `waitpid` to consume them. Basically, `r_debug_native_continue` will check for them and consume them accordingly. This also answers your another comment regarding the `linux_attach_single_pid` in which the thread always starts with a SIGSTOP after being attached. This make me realize that I could have done the removing signal part in here instead which will be included in my next commit. Thanks